WPA Beanstalk is crashing sometimes with  nil pointer error.

WPA is crashing and restarting sometimes with the nil pointer error

```
I0702 11:51:51.353487       1 controller.go:484] dev-call_otp_updater-crstg1 minWorkers=0, maxDisruptableWorkers=0
I0702 11:51:51.353505       1 controller.go:356] dev-call_otp_updater-crstg1: messages: 0, idle: 0, desired: 0
I0702 11:51:51.353513       1 controller.go:593] crstg1/25-29871-otpcallupdater-praccomm-api: WPA status is already up to date
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1202436]

goroutine 45213 [running]:
github.com/beanstalkd/go-beanstalk.(*Conn).cmd(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x16c2531, 0xa, 0xc00d553a38, 0x1, ...)
        /src/vendor/github.com/beanstalkd/go-beanstalk/conn.go:76 +0x66
github.com/beanstalkd/go-beanstalk.(*Tube).Stats(0xc00d553a90, 0x34, 0xc00085c1de, 0x16)
        /src/vendor/github.com/beanstalkd/go-beanstalk/tube.go:93 +0xcf
github.com/practo/k8s-worker-pod-autoscaler/pkg/queue.(*beanstalkClient).executeGetStats(0xc00734c440, 0xc00d553b98, 0x139935a, 0x1541140, 0x15dc740)
        /src/pkg/queue/beanstalk.go:123 +0x90
github.com/practo/k8s-worker-pod-autoscaler/pkg/queue.(*beanstalkClient).getStats(0xc00734c440, 0xc000a2ec00, 0x34, 0x1a0bec0, 0xc00734c440)
        /src/pkg/queue/beanstalk.go:144 +0x2f
github.com/practo/k8s-worker-pod-autoscaler/pkg/queue.(*Beanstalk).getMessages(0xc00017e090, 0xc000a2ec00, 0x34, 0x446235, 0xa1bc22, 0xc00e35a450)
        /src/pkg/queue/beanstalk.go:262 +0x8a
github.com/practo/k8s-worker-pod-autoscaler/pkg/queue.(*Beanstalk).poll(0xc00017e090, 0xc0014c3b00, 0x26, 0xc000a2ec1e, 0x16, 0xc00b599a70, 0x6, 0xc000a2ec00, 0x34, 0xc000a2ec0c, ...)
        /src/pkg/queue/beanstalk.go:351 +0x75
github.com/practo/k8s-worker-pod-autoscaler/pkg/queue.(*Poller).runPollThread(0xc00017e0f0, 0xc0014c3b00, 0x26)
        /src/pkg/queue/poller.go:46 +0x88
created by github.com/practo/k8s-worker-pod-autoscaler/pkg/queue.(*Poller).Run
        /src/pkg/queue/poller.go:93 +0x209
```

--- a/hack/format.sh
+++ b/hack/format.sh
@@ -1,7 +1,7 @@
 #!/bin/sh
 set -o errexit
 set -o nounset
-set -o pipefail
+#set -o pipefail
 
 find . -type f | grep .go$ | grep -v vendor | xargs -I {} gofmt -w {}
 

--- a/pkg/queue/beanstalk.go
+++ b/pkg/queue/beanstalk.go
@@ -2,6 +2,7 @@ package queue
 
 import (
 	"errors"
+	"fmt"
 	"io"
 	"net/url"
 	"path"

--- a/pkg/queue/beanstalk.go
+++ b/pkg/queue/beanstalk.go
@@ -325,7 +325,8 @@ func (b *Beanstalk) waitForShortPollInterval() {
 func (b *Beanstalk) reestablishConn(queueURI string) {
 	client, err := b.getClient(queueURI)
 	if err != nil {
-		klog.Error(err)
+		klog.Errorf("Could not reestablish conn, err:%v\n", err)
+		return
 	}
 
 	err = client.reestablishConn()

--- a/pkg/queue/beanstalk.go
+++ b/pkg/queue/beanstalk.go
@@ -61,6 +61,7 @@ type BeanstalkClientInterface interface {
 	put(body []byte, pri uint32, delay, t time.Duration) (id uint64, err error)
 	getStats() (int32, int32, int32, error)
 	longPollReceiveMessage(longPollInterval int64) (int32, int32, error)
+	reestablishConn() error
 }
 
 type beanstalkClient struct {
@@ -318,6 +319,16 @@ func (b *Beanstalk) waitForShortPollInterval() {
 	time.Sleep(b.shortPollInterval)
 }
 
+func (b *Beanstalk) reestablishConn(queueURI string) {
+	client, err := b.getClient(queueURI)
+	if err != nil {
+		klog.Error(err)
+	}
+
+	err = client.reestablishConn()
+	klog.Error(err)
+}
+
 func (b *Beanstalk) GetName() string {
 	return b.name
 }
@@ -335,6 +346,7 @@ func (b *Beanstalk) poll(key string, queueSpec QueueSpec) {
 		if err != nil {
 			klog.Errorf("Unable to perform request long polling %q, %v.",
 				queueSpec.name, err)
+			b.reestablishConn(queueSpec.uri)
 			return
 		}
 
@@ -360,6 +372,7 @@ func (b *Beanstalk) poll(key string, queueSpec QueueSpec) {
 	if err != nil {
 		klog.Errorf("Unable to get approximate messages in queue %q, %v.",
 			queueSpec.name, err)
+		b.reestablishConn(queueSpec.uri)
 		return
 	}
 	klog.V(3).Infof("%s: approxMessages=%d", queueSpec.name, approxMessages)
@@ -387,6 +400,7 @@ func (b *Beanstalk) poll(key string, queueSpec QueueSpec) {
 	if err != nil {
 		klog.Errorf("Unable to fetch idle workers %q, %v.",
 			queueSpec.name, err)
+		b.reestablishConn(queueSpec.uri)
 		time.Sleep(100 * time.Millisecond)
 		return
 	}

--- a/pkg/queue/beanstalk.go
+++ b/pkg/queue/beanstalk.go
@@ -264,6 +264,9 @@ func (b *Beanstalk) getClient(
 	if err != nil {
 		return nil, err
 	}
+	if client == nil {
+		return nil, fmt.Errorf("Not able to make client for: %s\n", queueURI)
+	}
 	b.clientPool.Store(queueURI, client)
 
 	return client.(BeanstalkClientInterface), nil

--- a/pkg/queue/beanstalk.go
+++ b/pkg/queue/beanstalk.go
@@ -92,6 +92,11 @@ func getBeanstalkConn(queueURI string) (*beanstalk.Conn, error) {
 	if err != nil {
 		return nil, errors.New("dial-error: " + err.Error())
 	}
+
+	if conn == nil {
+		return nil, fmt.Errorf("Connection nil for: %s\n", queueURI)
+	}
+
 	return conn, nil
 }
 

--- a/pkg/queue/beanstalk.go
+++ b/pkg/queue/beanstalk.go
@@ -196,8 +196,17 @@ func (c *beanstalkClient) put(
 func (c *beanstalkClient) doLongPoll(
 	longPollInterval int64) (bool, uint64, error) {
 
+	var id uint64
+	var err error
+
 	tubeSet := beanstalk.NewTubeSet(c.conn, path.Base(c.queueURI))
-	id, _, err := tubeSet.Reserve(
+	if tubeSet == nil {
+		err = c.reestablishConn()
+		if err != nil {
+			return false, id, err
+		}
+	}
+	id, _, err = tubeSet.Reserve(
 		time.Duration(longPollInterval) * time.Second)
 	if err == nil {
 		return true, id, nil

--- a/pkg/queue/beanstalk_mock.go
+++ b/pkg/queue/beanstalk_mock.go
@@ -84,3 +84,17 @@ func (mr *MockBeanstalkClientInterfaceMockRecorder) put(arg0, arg1, arg2, arg3 i
 	mr.mock.ctrl.T.Helper()
 	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "put", reflect.TypeOf((*MockBeanstalkClientInterface)(nil).put), arg0, arg1, arg2, arg3)
 }
+
+// reestablishConn mocks base method
+func (m *MockBeanstalkClientInterface) reestablishConn() error {
+	m.ctrl.T.Helper()
+	ret := m.ctrl.Call(m, "reestablishConn")
+	ret0, _ := ret[0].(error)
+	return ret0
+}
+
+// put indicates an expected call of put
+func (mr *MockBeanstalkClientInterfaceMockRecorder) reestablishConn() *gomock.Call {
+	mr.mock.ctrl.T.Helper()
+	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "reestablishConn", reflect.TypeOf((*MockBeanstalkClientInterface)(nil).reestablishConn))
+}

--- a/pkg/queue/sqs.go
+++ b/pkg/queue/sqs.go
@@ -325,7 +325,12 @@ func (s *SQS) getAverageNumberOfMessagesSent(queueURI string) (float64, error) {
 		},
 	}
 
-	result, err := s.getCWClient(queueURI).GetMetricData(&cloudwatch.GetMetricDataInput{
+	cwClient, err := s.getCWClient(queueURI)
+	if err != nil {
+		return 0.0, err
+	}
+
+	result, err := cwClient.GetMetricData(&cloudwatch.GetMetricDataInput{
 		EndTime:           &endTime,
 		StartTime:         &startTime,
 		MetricDataQueries: []*cloudwatch.MetricDataQuery{query},

--- a/pkg/queue/sqs.go
+++ b/pkg/queue/sqs.go
@@ -87,8 +87,13 @@ func (s *SQS) getSQSClient(queueURI string) *sqs.SQS {
 	return s.sqsClientPool[getRegion(queueURI)]
 }
 
-func (s *SQS) getCWClient(queueURI string) *cloudwatch.CloudWatch {
-	return s.cwClientPool[getRegion(queueURI)]
+func (s *SQS) getCWClient(queueURI string) (*cloudwatch.CloudWatch, error) {
+	client, ok := s.cwClientPool[getRegion(queueURI)]
+	if !ok {
+		return nil, fmt.Errorf("Client not found for queue: %s\n", queueURI)
+	}
+
+	return client, nil
 }
 
 func (s *SQS) longPollReceiveMessage(queueURI string) (int32, error) {
@@ -187,7 +192,12 @@ func (s *SQS) getNumberOfMessagesReceived(queueURI string) (float64, error) {
 		},
 	}
 
-	result, err := s.getCWClient(queueURI).GetMetricData(&cloudwatch.GetMetricDataInput{
+	cwClient, err := s.getCWClient(queueURI)
+	if err != nil {
+		return 0.0, err
+	}
+
+	result, err := cwClient.GetMetricData(&cloudwatch.GetMetricDataInput{
 		EndTime:           &endTime,
 		StartTime:         &startTime,
 		MetricDataQueries: []*cloudwatch.MetricDataQuery{query},

